### PR TITLE
Add object logging

### PR DESCRIPTION
**Pull request by matthiasbeyer**
This PR adds the object logging mechanism after fixing some interface defintions for this.

### CODE REVIEW

File: src/objects/object.h [hunk 1]
```diff
@@ -75,10 +77,11 @@ typedef bool (*ws_object_deinit_callback)(struct ws_object* const);
 
 /**
  * log callback
- *
- * @todo Implement second parameter: struct ws_log_context*
  */
-typedef bool (*ws_object_log_callback)(struct ws_object* const, void*);
+typedef bool (*ws_object_log_callback)(
+        struct ws_object* self,
+        struct ws_logger_context* const log_ctx
```
Comment (neithernut) [inline src/objects/object.h]: Context comes first.

**Comment by matthiasbeyer** [comment]
Would it be beneficial to add a function which does literally the same but _always_ on `stderr`, and only including this function when compiling with debugging? So you could replace the vtable pointer and every call to `ws_object_log()` would cause debugging information.

A function for preventing a logging call would be nice, too (for debugging, of course). Something like `debug_dummy` or so.

What are you guys thinking about this? The overhead in the tests would be minimal, as you would only have to override the vtable ptr in the preface (is this actually possible when having constant vtables?)

**Comment by neithernut** [comment]
Tests can ship replacement functions for logging anyways...

**Comment by matthiasbeyer** [comment]
Indeed, so no need to add more code here.

**Comment by neithernut** [comment]
Indeed.

**Comment by neithernut** [comment]
Mh... do we have actually a usecase for this?

**Comment by matthiasbeyer** [comment]
Mainly debugging.

**Comment by matthiasbeyer** [comment]
Umm... but also normal logging. Something like "log that this monitor has changed, log that these windows were moved" etc.

**Comment by neithernut** [comment]
I don't see how a logging function in _each_ object type would help. Printing the object pointer, hash and type would be sufficient. And that should be done using a regular logging call, not one embedded in the object model.

**Comment by matthiasbeyer** [comment]
The callback in the object type is actually for doing deep-logging, printing attributes and so on. In case of a moved window this could be the new position for example.

**Comment by neithernut** [comment]
So we dump the entire state of the object? Urgh...
In this case the name should be changed to `dump_state` or something.

**Comment by matthiasbeyer** [comment]
A rename would break the build in various places, I guess.

**Comment by neithernut** [comment]
I seriously doubt that. The only derived types are still in PRs.

**Comment by matthiasbeyer** [comment]
Btw, @neithernut : The rename crashed the build as expected.

Please review again. And merge this if acknowledged.

**Comment by neithernut** [comment]
ACK.

**Pull request merged**